src/Browser/Avalonia.Browser/webapp/modules/avalonia/rendering/softwareRenderTarget.ts

```diff
@@ -0,0 +1,44 @@
+import { RuntimeAPI } from "../../../types/dotnet";
+import { WebRenderTarget } from "./webRenderTarget";
+
+export class SoftwareRenderTarget extends WebRenderTarget {
+    private readonly runtime: RuntimeAPI | undefined;
+    private readonly context: CanvasRenderingContext2D | OffscreenCanvasRenderingContext2D;
+    constructor(canvas: HTMLCanvasElement | OffscreenCanvas) {
+        const context = canvas.getContext("2d", {
+            alpha: true
+        });
+        if (!context) {
+            throw new Error("HTMLCanvasElement.getContext(2d) returned null.");
+        }
+
+        super(canvas, "software");
+        this.context = context;
+
+        this.runtime = globalThis.getDotnetRuntime(0);
+    }
+
+    public putPixelData(pointer: number, length: number, width: number, height: number): void {
+        const heap8 = this.runtime?.localHeapViewU8();
```

Comment:
This code assumes, that .net runtime on wasm will always put our buffer into this specific heap. Previously there was a fallback with IMemoryView copy.

But we should be fine with that at least for .NET 8+9.

Reply:
We are now using `Marshal.AllocHGlobal` instead of managed memory, since RetainedFramebuffer class is now in use.